### PR TITLE
Add external data store configuration HZ-1144

### PR DESCRIPTION
**Pull request by ldziedziul**
Adds initial support for external data store configuration:

Related to https://hazelcast.atlassian.net/browse/HZ-1144

- Adds `external-data-store` configuration section to hazelcast config files (XML/YAML).
- Adds `ExternalDataStoreService` to `NodeEngine` for accessing configured data stores
- Adds simplistic implementation of `JdbcDataStoreFactory` for handling JDBC data stores
- Reusing datastore resources (using the `shared` flag) is not yet implemented, will be addressed later within https://hazelcast.atlassian.net/browse/HZ-1145
- Adds support for referencing external datastore in Sources.jdbc:
```java
Sources.jdbc(externalDataStoreRef("mysql-database"), "SELECT * ..", (rs) -> ...)
```

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    private String className;
+    private boolean shared;
+    private Properties properties = new Properties();
+
+    public ExternalDataStoreConfig() {
+    }
+
+    @Override
+    public ExternalDataStoreConfig setName(String name) {
+        this.name = checkNotNull(name, "Name must not be null");
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java:41]: I like you added human-friendly message, not just "name" as it is in many places :)

File: hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import static com.hazelcast.internal.util.Preconditions.checkHasText;
+import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+
+/**
+ * Contains configuration for external data store
+ */
+public class ExternalDataStoreConfig implements NamedConfig {
+
+    private String name;
+    private String className;
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java:32]: Can't we store just Class<?> ?

File: hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+     * should be created
+     *
+     * @return if the data store instance should be reused
+     */
+    public boolean isShared() {
+        return shared;
+    }
+
+    /**
+     * true if an instance of the external data store will be reused. false when on each usage the data store instance
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java:80]: nit: {@code true}

File: hazelcast/src/main/java/com/hazelcast/datastore/JdbcDataStoreFactory.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+public class JdbcDataStoreFactory implements ExternalDataStoreFactory<DataSource> {
+    private ExternalDataStoreConfig config;
+
+    public void init(ExternalDataStoreConfig config) {
+        this.config = config;
+    }
+
+    @Override
+    public DataSource createDataStore() {
+        String jdbcUrl = config.getProperties().getProperty("jdbc.url");
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/datastore/JdbcDataStoreFactory.java:33]: maybe add helper method: config.getProperty()

File: hazelcast/src/main/java/com/hazelcast/datastore/impl/package-info.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * <p>This package contains implementation handling external data stores.<br>
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/datastore/impl/package-info.java:18]:
```suggestion
 * <p>This package contains implementation of handling external data stores.<br>
```

File: hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java (hunk 2 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java:32]: I followed approach from https://github.com/hazelcast/hazelcast/blob/be07ca930186e8140f8c5fb8e6afbe60a7cadc92/hazelcast/src/main/java/com/hazelcast/config/MapStoreConfig.java#L56

File: hazelcast/src/main/java/com/hazelcast/datastore/JdbcDataStoreFactory.java (hunk 4 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/datastore/JdbcDataStoreFactory.java]: fixed: https://github.com/hazelcast/hazelcast/pull/21716/commits/284de25d4cfa21f89e12a9d4815692d3ebd9d35d

File: hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java (hunk 3 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/config/ExternalDataStoreConfig.java]: fixed: https://github.com/hazelcast/hazelcast/pull/21716/commits/896d572f722ebb1d26a12634351385f253fdbed3

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadJdbcP.java [hunk 6]
```diff
@@ -77,6 +78,17 @@ public static <T> ProcessorMetaSupplier supplier(
             @Nonnull SupplierEx<? extends Connection> newConnectionFn,
             @Nonnull ToResultSetFunction resultSetFn,
             @Nonnull FunctionEx<? super ResultSet, ? extends T> mapOutputFn
+    ) {
+        return supplier(ctx -> newConnectionFn.get(), resultSetFn, mapOutputFn);
+    }
+
+    /**
+     * Use {@link SourceProcessors#readJdbcP}.
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadJdbcP.java:86]: typo: Used by

**Review by TomaszGaweda — APPROVED** [review]
Looks good, only minor comments to consider

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadJdbcP.java (hunk 6 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadJdbcP.java:86]: I think it's not a typo

File: hazelcast/src/main/java/com/hazelcast/jet/pipeline/ExternalDataStoreRef.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.function.ToResultSetFunction;
+
+import java.io.Serializable;
+
+/**
+ * Represents a reference to the external datastore, used with {@link Sources#jdbc(
+ *ExternalDataStoreRef, ToResultSetFunction, FunctionEx)}.
+ */
+public class ExternalDataStoreRef implements Serializable {
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/pipeline/ExternalDataStoreRef.java:28]: Yes,  less stringly-typed.

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 8]
```diff
@@ -3092,6 +3095,42 @@ public Config setIntegrityCheckerConfig(final IntegrityCheckerConfig integrityCh
         return this;
     }
 
+    /**
+     * Returns the map of external data store configurations, mapped by config name.
+     *
+     * @since 5.2
+     */
+    public Map<String, ExternalDataStoreConfig> getExternalDataStoreConfigs() {
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java:3103]: I just followed the approach from other Config maps

File: hazelcast/src/test/java/com/hazelcast/datastore/impl/ExternalDataStoreServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    private final Config config = new Config();
+
+    @Before
+    public void configure() {
+        Properties properties = new Properties();
+        properties.put("jdbc.url", "jdbc:h2:mem:" + ExternalDataStoreServiceImplTest.class.getSimpleName());
+        ExternalDataStoreConfig externalDataStoreConfig = new ExternalDataStoreConfig()
+                .setName("test-data-store")
+                .setClassName("com.hazelcast.datastore.JdbcDataStoreFactory")
```
Comment (frant-hartm) [inline hazelcast/src/test/java/com/hazelcast/datastore/impl/ExternalDataStoreServiceImplTest.java:55]: Add tests for
- class which doesn't exist
- class which doesn't implement the `ExternalDataStoreFactory` interface

File: hazelcast/src/test/java/com/hazelcast/jet/impl/connector/WriteJdbcPTest.java [hunk 10]
```diff
@@ -86,35 +96,52 @@ public void setup() throws SQLException {
...
-                 }
-         ));
+                .map(item -> entry(item, item.toString()))
+                .writeTo(Sinks.jdbc("INSERT INTO " + tableName + " VALUES(?, ?)",
+                        () -> createDataSource(false),
+                        (stmt, item) -> {
+                            stmt.setInt(1, item.getKey());
+                            stmt.setString(2, item.getValue());
+                        }
+                ));
```
Comment (frant-hartm) [inline hazelcast/src/test/java/com/hazelcast/jet/impl/connector/WriteJdbcPTest.java:114]: nit: Incorrect indentation?

**Review by frant-hartm — CHANGES_REQUESTED** [review]
Nice work. 

Let's mark all new configs and public APIs with `@Beta` for 5.2 (I commented where appropriate, but might have forgotten something). It will leave us room for change if we need to.

I will integrate it with the SQL connector to verify it works as expected.

**Comment by ldziedziul** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/client/impl/clientside/ClientDynamicClusterConfig.java [hunk 11]
```diff
@@ -1112,6 +1114,22 @@ public Config setIntegrityCheckerConfig(final IntegrityCheckerConfig config) {
         throw new UnsupportedOperationException(UNSUPPORTED_ERROR_MESSAGE);
     }
 
+    @Override
+    public Map<String, ExternalDataStoreConfig> getExternalDataStoreConfigs() {
+        throw new UnsupportedOperationException(UNSUPPORTED_ERROR_MESSAGE);
+    }
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/client/impl/clientside/ClientDynamicClusterConfig.java:1120]: Yes I will do it separately

File: hazelcast/src/main/java/com/hazelcast/jet/core/processor/SourceProcessors.java [hunk 12]
```diff
@@ -429,14 +435,41 @@ public static <T> ProcessorMetaSupplier streamJmsTopicP(
...
+     *
+     * @since 5.2
+     */
+    public static <T> ProcessorMetaSupplier readJdbcP(
+            @Nonnull ExternalDataStoreRef externalDataStoreRef,
+            @Nonnull ToResultSetFunction resultSetFn,
+            @Nonnull FunctionEx<? super ResultSet, ? extends T> mapOutputFn
+    ) {
+        return ReadJdbcP.supplier(context ->
+                        getDataStoreFactory(context, externalDataStoreRef.getName()).createDataStore().getConnection(),
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/core/processor/SourceProcessors.java:460]: Maybe we can rename `createDataStore()` to `getDataStore()` and return the same instance when `shared` is enabled?

[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/core/processor/SourceProcessors.java]: Fixed by getting DataSource in the ProcessorSupplier and sharing it between processors of a job:  https://github.com/hazelcast/hazelcast/pull/21716/commits/66dd768ff0e54844cddc13828675fb970d1499eb

File: hazelcast/src/main/java/com/hazelcast/datastore/JdbcDataStoreFactory.java [hunk 13]
```diff
@@ -33,23 +34,26 @@ public class JdbcDataStoreFactory implements ExternalDataStoreFactory<DataSource
...
-            throw new UnsupportedOperationException("shared flag is not yet supported");
-        }
-        if (config.getProperties().containsKey(JDBC_USERNAME) || config.getProperties().containsKey(JDBC_PASSWORD)) {
-            String username = config.getProperty(JDBC_USERNAME);
-            String password = config.getProperty(JDBC_PASSWORD);
-            return new DataSourceFromJdbcUrl(jdbcUrl, username, password);
-        }
-        //TODO Use HikariDataSource for pooling or similar
-        return new DataSourceFromJdbcUrl(jdbcUrl);
+        HikariConfig dataSourceConfig = new HikariConfig();
```
Comment (ldziedziul) [inline hazelcast/src/main/java/com/hazelcast/datastore/JdbcDataStoreFactory.java:53]: Hikari's default pool size is 10, do we want to expose more internal settings of data sources to the users? (see https://github.com/brettwooldridge/HikariCP#gear-configuration-knobs-baby)?

**Comment by ldziedziul** [comment]
run-lab-run

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/jet/core/processor/SourceProcessors.java [hunk 14]
```diff
@@ -429,14 +436,40 @@ public static <T> ProcessorMetaSupplier streamJmsTopicP(
 
     /**
      * Returns a supplier of processors for {@link Sources#jdbc(
-     * SupplierEx, ToResultSetFunction, FunctionEx)}.
+     *SupplierEx, ToResultSetFunction, FunctionEx)}.
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/core/processor/SourceProcessors.java:439]: It's the default formatting of the IntelliJ, I would prefer and keep it as it is. WDYT?

File: hazelcast/pom.xml [hunk 15]
```diff
@@ -748,5 +762,11 @@
             <version>${project.parent.version}</version>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.zaxxer</groupId>
+            <artifactId>HikariCP</artifactId>
+            <version>${hikari.version}</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/pom.xml:770]: Hmm, sure we could split, but we would have basically a module with a single class and a single dependency, are we fine with that?

File: hazelcast/src/main/java/com/hazelcast/jet/pipeline/JdbcSinkBuilder.java [hunk 16]
```diff
@@ -183,15 +189,39 @@ public JdbcSinkBuilder<T> batchLimit(int batchLimit) {
...
+     * Sets the reference to the configured external dataStore of {@link ExternalDataStoreRef} from which
+     * the instance of the {@link javax.sql.DataSource} will be retrieved.
+     *
+     * @param externalDataStoreRef the reference to the configured external dataStore
+     * @return this instance for fluent API
+     * @since 5.2
+     */
+    @Nonnull
+    @Beta
+    public JdbcSinkBuilder<T> externalDataStoreRef(ExternalDataStoreRef externalDataStoreRef) {
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/pipeline/JdbcSinkBuilder.java]: fixed: https://github.com/hazelcast/hazelcast/pull/21716/commits/7084cd03a4bb414889f7da6cb31724668b29f67b

File: hazelcast/src/main/java/com/hazelcast/jet/pipeline/Sources.java [hunk 17]
```diff
@@ -1328,6 +1328,41 @@ public static <T> BatchSource<T> jdbc(
...
+    /**
+     * Returns a source which connects to the specified database using the given
+     * {@code externalDataStoreRef}, queries the database and creates a result set
+     * using the the given {@code resultSetFn}. It creates output objects from the
+     * {@link ResultSet} using given {@code mapOutputFn} and emits them to
+     * downstream.
+     * <p>
+     * Example: <pre>{@code
+     *     p.readFrom(Sources.jdbc(
+     *         ExternalDataStoreRef.externalDataStoreRef(JDBC_DATA_STORE),
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/pipeline/Sources.java]: Done: https://github.com/hazelcast/hazelcast/pull/21716/commits/458d764188d0942ef39a88f5410b22ff418fab8f

File: hazelcast/src/main/java/com/hazelcast/jet/pipeline/JdbcSinkBuilder.java [hunk 18]
```diff
@@ -183,15 +189,39 @@ public JdbcSinkBuilder<T> batchLimit(int batchLimit) {
...
+
+    }
+
+    /**
+     * Sets the reference to the configured external dataStore of {@link ExternalDataStoreRef} from which
+     * the instance of the {@link javax.sql.DataSource} will be retrieved.
+     *
+     * @param externalDataStoreRef the reference to the configured external dataStore
+     * @return this instance for fluent API
+     * @since 5.2
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/jet/pipeline/JdbcSinkBuilder.java]: Done: https://github.com/hazelcast/hazelcast/pull/21716/commits/e6ba5188d08b7a9e446c5cd38923396dfe4212de

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 19]
```diff
@@ -3092,6 +3096,56 @@ public Config setIntegrityCheckerConfig(final IntegrityCheckerConfig integrityCh
...
+     * @since 5.2
+     */
+    @Beta
+    public Map<String, ExternalDataStoreConfig> getExternalDataStoreConfigs() {
+        return externalDataStoreConfigs;
+    }
+
+    /**
+     * Sets the map of external data store configurations, mapped by config name.
+     *
```
[the remark this answers is not in the document]
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java]: Done: https://github.com/hazelcast/hazelcast/pull/21716/commits/e6ba5188d08b7a9e446c5cd38923396dfe4212de

**Comment by ldziedziul** [comment]
> I thought we would pass the properties configured for the data store:

Makes sense, done: https://github.com/hazelcast/hazelcast/pull/21716/commits/e0c440d357046672e1bce487720e58a07a576e1c

**Comment by ldziedziul** [comment]
run-lab-run

File: hazelcast/pom.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast/pom.xml:770]: Let's leave it where it is for now.

**Review by frant-hartm — APPROVED** [review]
All good. Thanks!

**Review by mdumandag — COMMENTED** [review]
I quickly went over the checklist in https://github.com/hazelcast/hazelcast/wiki/Checklist-for-Hazelcast-config-changes, and I believe there are some missing items.

- Do we have spring support for this? If so, items related to it are not done
- We should add tests for YAML schema validation
- We should add this to hazelcast-fullconfig-without-network.[xml&yaml]
- ConfigXmlGenerator support seems missing
- ConfigCompatibilityChecker support seems missing
- Should we have readonly versions of this config, per this item `add or update read-only config in com.hazelcast.internal.config.*ReadOnly`

**Comment by ldziedziul** [comment]
Many thanks for the review!
 
> * Do we have spring support for this? If so, items related to it are not done

Will be added in a separate PR
> * We should add tests for YAML schema validation

Added https://github.com/hazelcast/hazelcast/pull/21716/commits/e618c84e1d1fedf33d60d4bff1a2f3d2503cb728
> * We should add this to hazelcast-fullconfig-without-network.[xml&yaml]

Added https://github.com/hazelcast/hazelcast/pull/21716/commits/05303a4cba9574d0de4bac2b8c2244c59c60e400
> * ConfigXmlGenerator support seems missing

Added https://github.com/hazelcast/hazelcast/pull/21716/commits/0ca6fdab83de96122dafa39ea721fcac408c3aa7
> * ConfigCompatibilityChecker support seems missing

Added https://github.com/hazelcast/hazelcast/pull/21716/commits/d31e71318eb0394fea4c634f71e7551fae083f7d
> * Should we have readonly versions of this config, per this item `add or update read-only config in com.hazelcast.internal.config.*ReadOnly`

Added https://github.com/hazelcast/hazelcast/pull/21716/commits/8f49f3dee72f3d337c63b61e5b5ee0d42390d589

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java [hunk 20]
```diff
@@ -660,6 +662,28 @@ public static void pnCounterYamlGenerator(Map<String, Object> parent, Config con
         parent.put("pn-counter", child);
     }
 
+    public static void externalDataStoreYamlGenerator(Map<String, Object> parent, Config config) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:665]: Minor: Maybe we can do this(and DynamicConfigYamlGeneratorTest) later when you send the dynamic configuration PR?

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 21]
```diff
@@ -3092,6 +3097,125 @@ public Config setIntegrityCheckerConfig(final IntegrityCheckerConfig integrityCh
...
+     * @see #getConfigPatternMatcher()
+     * @see EvictionConfig#setSize(int)
+     * @since 5.2
+     */
+    @Beta
+    public ExternalDataStoreConfig findExternalDataStoreConfig(String name) {
+        name = getBaseName(name);
+        ExternalDataStoreConfig config = lookupByPattern(configPatternMatcher, externalDataStoreConfigs, name);
+        if (config != null) {
+            return new ExternalDataStoreConfig(config);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/config/Config.java:3214]: This should be  `ExternalDataStoreConfigReadOnly`

**Review by mdumandag — COMMENTED** [review]
Changes look good to me overall. Left two (one minor) comment regarding the recent changes. After getting them resolved, I can approve the PR

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java (hunk 20 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:665]: I'll create the PR right after this one being merged, if you don't mind I would leave it as it is

File: hazelcast/src/main/java/com/hazelcast/config/Config.java (hunk 21 above)
Comment (ldziedziul) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java]: Fixed: https://github.com/hazelcast/hazelcast/pull/21716/commits/65c88af16d3c43454a4e538c71e6420d31fd7ae4

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java (hunk 20 above)
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:665]: Ok, sounds good to me

**Review by mdumandag — APPROVED** [review]

**Pull request merged**